osu.Game.Tests/NonVisual/DifficultyAdjustmentModCombinationsTest.cs

```diff
@@ -20,124 +22,136 @@ public void TestNoMods()
-            Assert.IsTrue(((MultiMod)combinations[2]).Mods[1] is ModB);
-            Assert.IsTrue(((MultiMod)combinations[2]).Mods[2] is ModC);
-            Assert.IsTrue(((MultiMod)combinations[3]).Mods[0] is ModB);
-            Assert.IsTrue(((MultiMod)combinations[3]).Mods[1] is ModC);
+            assertCombinations(new[]
+            {
+                new[] { typeof(ModNoMod) },
+                new[] { typeof(ModA) },
+                new[] { typeof(ModA), typeof(ModB), typeof(ModC) },
+                new[] { typeof(ModB), typeof(ModC) }
+            }, combinations);
         }
 
         [Test]
         public void TestIncompatibleThroughMultiMod()
         {
             var combinations = new TestLegacyDifficultyCalculator(new ModA(), new MultiMod(new ModB(), new ModIncompatibleWithA())).CreateDifficultyAdjustmentModCombinations();
 
-            Assert.AreEqual(3, combinations.Length);
-            Assert.IsTrue(combinations[0] is ModNoMod);
-            Assert.IsTrue(combinations[1] is ModA);
-            Assert.IsTrue(combinations[2] is MultiMod);
-
-            Assert.IsTrue(((MultiMod)combinations[2]).Mods[0] is ModB);
-            Assert.IsTrue(((MultiMod)combinations[2]).Mods[1] is ModIncompatibleWithA);
+            assertCombinations(new[]
+            {
+                new[] { typeof(ModNoMod) },
+                new[] { typeof(ModA) },
+                new[] { typeof(ModB), typeof(ModIncompatibleWithA) }
+            }, combinations);
         }
 
         [Test]
         public void TestIncompatibleWithSameInstanceViaMultiMod()
         {
             var combinations = new TestLegacyDifficultyCalculator(new ModA(), new MultiMod(new ModA(), new ModB())).CreateDifficultyAdjustmentModCombinations();
 
-            Assert.AreEqual(3, combinations.Length);
-            Assert.IsTrue(combinations[0] is ModNoMod);
-            Assert.IsTrue(combinations[1] is ModA);
-            Assert.IsTrue(combinations[2] is MultiMod);
+            assertCombinations(new[]
+            {
+                new[] { typeof(ModNoMod) },
+                new[] { typeof(ModA) },
+                new[] { typeof(ModA), typeof(ModB) }
+            }, combinations);
+        }
 
-            Assert.IsTrue(((MultiMod)combinations[2]).Mods[0] is ModA);
-            Assert.IsTrue(((MultiMod)combinations[2]).Mods[1] is ModB);
+        private void assertCombinations(Type[][] expectedCombinations, Mod[] actualCombinations)
+        {
+            Assert.AreEqual(expectedCombinations.Length, actualCombinations.Length);
+
+            foreach (Type[] expected in expectedCombinations)
+            {
+                Type[] actualTypes = ModUtils.FlattenMods(actualCombinations).Select(m => m.GetType()).ToArray();
+
+                Assert.Multiple(() =>
+                {
+                    foreach (var expectedType in expected)
+                        Assert.Contains(expectedType, actualTypes);
+                });
+            }
```

Comment:
I don't know if it's just me, but the way this is written is confusing me, for two reasons. First off, `actualTypes` is computed in the inner loop, even though its value actually is the same in every iteration of the loop. Worse yet, the flattening of all `actualCombinations` and the containment check weakens the test.

Consider `TestIncompatibleWithSameInstanceViaMultiMod()`: the expected combinations are:
```
expectedCombinations = {{ModNoMod}, {ModA}, {ModA, ModB}}
```
But the `FlattenMods` operation returns a flat list instead, yielding
```
actualTypes = {ModNoMod, ModA, ModA, ModB}
```
And because the containment assert is checking each individual mod from the expected combination at once rather than the entire set, you can get false positives. As in, imagine that the last element in `expectedCombinations` was `{ModB}`, rather than `{ModA, ModB}`; `actualTypes` would still contain `{ModA}`, even though the behaviour would be erroneous. And conversely, if you change `{ModA, ModB}` in `expectedCombinations` to just `{ModB}`, the test passes erroneously for the same reason.

If I were to write this method myself, I would write:

```csharp
private void assertCombinations(Type[][] expectedCombinations, Mod[] actualCombinations)
{
    Assert.AreEqual(expectedCombinations.Length, actualCombinations.Length);

    Assert.Multiple(() =>
    {
        for (int i = 0; i < expectedCombinations.Length; ++i)
        {
            Type[] expectedTypes = expectedCombinations[i];
            Type[] actualTypes = ModUtils.FlattenMod(actualCombinations[i]).Select(m => m.GetType()).ToArray();

            Assert.That(expectedTypes, Is.EquivalentTo(actualTypes));
        }
    });
}
```

which would correctly fail in the cases I gave, because entire sets of mods are being checked, rather than only individual parts of these sets.

Reply:
Good catch, I've applied your fix.